src/kmod/ktfutce.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2018-2020 VMware, Inc.
+ * SPDX-License-Identifier: GPL-2.0-or-later
+ */
+
+#include "priv.h"
+#include "process-tracking.h"
+#include <linux/kthread.h>
+#include <linux/wait.h>
+
+//
+// ktfutce handles a queue of Fork events and the tgid in question.
+// We may want to pre-allocate a process tracking table entry and
+// send to here as well and fill in all the data.
+//
+// With our current fork hook we cannot safely lookup the new
+// task yet. So we throw the job into our kthread via a
+// wait queue. By the time this thread wakes the new task
+// will highly likely be available.
+//
+
+static DECLARE_WAIT_QUEUE_HEAD(clone_wait);
+struct clone_struct {
+	struct task_struct *task;
+	struct list_head    head;
+	spinlock_t	    lock;
+	unsigned long	    len;
+};
+static struct clone_struct clone_queue = { .task = NULL };
+
+struct clone_event {
+	struct list_head list;
+	pid_t		 pid;
+	struct CB_EVENT *event;
+};
+
+static int ktfutce_thread(void *arg);
+
+int ktfutce_register(void)
+{
+	int ret = 0;
+
+	memset(&clone_queue, 0, sizeof(clone_queue));
+	INIT_LIST_HEAD(&clone_queue.head);
+	spin_lock_init(&clone_queue.lock);
+
+	clone_queue.task = kthread_run(ktfutce_thread, NULL, "cb_ktfutce");
+	if (IS_ERR(clone_queue.task)) {
+		ret		 = PTR_ERR(clone_queue.task);
+		clone_queue.task = NULL;
+		return ret;
+	}
+	return ret;
+}
+
+// Probably should be called before logger shutdown
+void ktfutce_shutdown(void)
+{
+	if (clone_queue.task) {
+		struct task_struct *task = clone_queue.task;
+		clone_queue.task	 = NULL;
+		wake_up_interruptible(&clone_wait);
+		kthread_stop(task);
+	}
+}
+
+// Enqueue a fork event and ask the kthread to work on it
+int ktfutce_add_pid(pid_t pid, struct CB_EVENT *event, gfp_t mode)
+{
+	struct clone_event *clone_event = NULL;
+
+	if (!clone_queue.task) {
+		return -EINVAL;
+	}
+
+	// Can't be zero!
+	if (!pid) {
+		return -EINVAL;
+	}
+
+	clone_event = kzalloc(sizeof(*clone_event), mode);
```

Comment:
ktfutce_add_pid allocates memory, where does it get freed?

Reply:
After clone_shift is called in `ktfutce_thread`. On shutdown we could try to iterate through the queue list and free things but I don't see that being a huge concern given the way our module has to be unloaded.